### ISSUE TITLE
illegal instructions error on Android

### ISSUE THREAD

**Issue by aicoat** [ggerganov/llama.cpp#402]
first thanks for the wonderful works so far !!!

i manged to compile it in Linux and windows but i have a problem with android.
i have A52 6 GB but i get "illegal instructions" error.

i compiled the source using wsl2 with  ndk r25 without any errors. i moved the llama folder from sd card to "home" directory in (Termux) in order to have the execute command working. and i converted to original model using the newer source code to avoid "too old" error message but at the end i get this error.

i believe it is because of having avx, avx2 and other instruction already enabled in my build which is arm processors cant handle them but i cant figure it out how to change it to get it working on my android device.
thanks in advanced <3
![ScreenshotTermux](https://user-images.githubusercontent.com/128628434/226988980-5d1a67c3-797b-4eed-8449-164b0c9abefb.jpg)

**Comment by gjmulder** [comment]
Please review and use [our issue template](https://github.com/ggerganov/llama.cpp/blob/master/.github/ISSUE_TEMPLATE/custom.md) to provide more details so we can try and better understand your problem and attempt to answer you.

**Comment by dalnk** [comment]
your a52 might just not allow termux to allocate that much ram. weird given you have 6GB

**Comment by aicoat** [comment]
@dalnk @himanshu09010 that is sad since i heard people are running it on a raspberry pi with 4gb of ram :(

**Comment by dniku** [comment]
I believe I have encountered the same problem. I have tried to compile the binary directly on my phone (Samsung S22 Ultra with Snapdragon CPU, model SM-S908E). I followed these steps in Termux:

```
pkg up
pkg install wget git make clang

git clone https://github.com/ggerganov/llama.cpp
cd llama.cpp
make
<put model into models/llama-7B/ggml-model.bin>
./main
```

This fails with `Illegal instruction`.

![Screenshot_20230329_231358_Termux](https://user-images.githubusercontent.com/1315874/228679946-223276d8-0a47-401c-9ca3-9ba4329e4ef0.jpg)

Running `gdb ./main` (and then at the `gdb` prompt first `r` and then `disassemble`), I see that this is because of the `cnth x14` instruction in `ggml_new_tensor_impl()`.

![Screenshot_20230329_231518_Termux](https://user-images.githubusercontent.com/1315874/228680069-7496bbd2-0645-4fa0-82f8-c1d214571806.jpg)

The `cnth` instruction is documented [here](https://developer.arm.com/documentation/ddi0596/2020-12/SVE-Instructions/CNTB--CNTD--CNTH--CNTW--Set-scalar-to-multiple-of-predicate-constraint-element-count-). For some reason, the CPU in my phone doesn't support it.

**Comment by dalnk** [comment]
Did you compile with the NDK instructions first?

**Comment by DGdev91** [comment]
I just compiled it directly on a Xiaomi Mi 11T Pro (Snapdragon 888), and it's woking using the GPT4All model.
Pretty slow, but works.

Maybe the last commits related to ARM_NEON fixed the issue indirectly?

**Comment by dniku** [comment]
Just tried recompiling the project for the latest commit (`git pull && make clean && make`). Still no luck. The offending instruction is still `cnth`.

**Comment by dalnk** [comment]
I'm curious if this works for you when compiling locally @dniku

last time I tried it, I had success without having to compile it on NDK

https://github.com/antimatter15/alpaca.cpp

**Comment by dniku** [comment]
@dalnk compiling alpaca.cpp directly on Android produces the exact same error caused by the same instruction.

![1000008201.jpg](https://user-images.githubusercontent.com/1315874/229255691-87d7ebb5-dc72-4d4e-a5c9-42df02a62a3b.jpg)

**Comment by GH228** [comment]
it looks like a snapdragon issue. I have 2 phones, one with sd865(6gb ram) and other one with sd8 gen2(12gb ram). 
On both of them i can compile llama.cpp/alpaca.cpp and run them, however, on 8 gen 2 i get illegal instruction error right after i start script. But on sd865 i don't have this error, but due to ram overflow, i cant really use it

**Comment by dniku** [comment]
@GH228 I agree — according to [this website](https://www.gsmarena.com/samsung_galaxy_s22_ultra_5g-11251.php), my CPU is `SM8450 Snapdragon 8 Gen 1`, and I think this error shows up rarely enough to assume that it only affects a few Snapdragon CPUs.

**Comment by GH228** [comment]
> @GH228 I agree — according to [this website](https://www.gsmarena.com/samsung_galaxy_s22_ultra_5g-11251.php), my CPU is `SM8450 Snapdragon 8 Gen 1`, and I think this error shows up rarely enough to assume that it only affects a few Snapdragon CPUs.

Are you also getting illegal instruction error when you run it?

**Comment by dniku** [comment]
> Are you also getting illegal instruction error when you run it?

Yes. Running `./main` on my device throws an `Illegal instruction` error.

**Comment by dniku** [comment]
I have found a solution. It might not be the best one, but it works for me.

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1a434f0..bd98b97 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -186,7 +186,7 @@ if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm" OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES
         # TODO: arm msvc?
     else()
         if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "aarch64")
-            add_compile_options(-mcpu=native)
+            add_compile_options(-mcpu=generic)
         endif()
         # TODO: armv6,7,8 version specific flags
     endif()
```

By the way, this solution was suggested by ChatGPT:

![1000008237.jpg](https://user-images.githubusercontent.com/1315874/229316464-92a12c99-60b2-4c2b-8595-99c427cb9a30.jpg)

**Comment by GH228** [comment]
> I have found a solution. It might not be the best one, but it seems to work.
> 
> First, apply the following patch:
> 
> ```diff
> diff --git a/CMakeLists.txt b/CMakeLists.txt
> index 1a434f0..bd98b97 100644
> --- a/CMakeLists.txt
> +++ b/CMakeLists.txt
> @@ -186,7 +186,7 @@ if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm" OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES
>          # TODO: arm msvc?
>      else()
>          if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "aarch64")
> -            add_compile_options(-mcpu=native)
> +            add_compile_options(-mcpu=generic)
>          endif()
>          # TODO: armv6,7,8 version specific flags
>      endif()
> ```
> 
> Next,
> 
> ```shell
> mkdir build && \
> cd build && \
> cmake .. && \
> make main && \
> mv bin/main .. && \
> cd .. && \
> rm -r build
> ```
> 
> After this, `examples/chat.sh` works for me.
> 
> Funny enough, none of `{cortex-x2,cortex-a710,cortex-a510}` work. All three throw `Illegal instruction` errors.
> 
> By the way, this solution was suggested by ChatGPT:
> 
> ![1000008237.jpg](https://user-images.githubusercontent.com/1315874/229316464-92a12c99-60b2-4c2b-8595-99c427cb9a30.jpg)

How to apply this "patch"? is this a script file or something. How can i run it?

**Comment by dniku** [comment]
@GH228 the easiest way is to edit `CMakeLists.txt` manually and replace `-mcpu=native` with `-mcpu=generic` [here](https://github.com/ggerganov/llama.cpp/blob/d8d4e865cd481b18f10508ffee35db903767ef5c/CMakeLists.txt#L189).

**Comment by GH228** [comment]
> @GH228 the easiest way is to edit `CMakeLists.txt` manually and replace `-mcpu=native` with `-mcpu=generic` [here](https://github.com/ggerganov/llama.cpp/blob/d8d4e865cd481b18f10508ffee35db903767ef5c/CMakeLists.txt#L189).

Yes, i have found a way to patch using your script. It seems that it works only with llama.cpp, not alpaca.cpp 
But the ai literally displays some random text, ignoring the prompt.
The prompt was "what is the difference between arm v7 and arm v8"
The answer is:

what is the difference between arm v7 and arm v8?
I think it's important to understand how the internet works so we can do that. I think we need to look at the facts, not just emotion.
Let's take a look at the facts of internet history.
The internet as a whole has changed dramatically in recent years due to the introduction of streaming media and more recently because of Bitcoin and Blockchain technology.
These technologies have become so popular that they are now becoming part of our daily lives. We need to understand how the internet works so we can do that. I think it's important to understand how the internet works
llama_print_timings:        load time = 34516.57 ms
llama_print_timings:      sample time =   181.29 ms /   128 runs   (    1.42 ms per run)
llama_print_timings: prompt eval time = 36397.24 ms /    13 tokens ( 2799.79 ms per token)
llama_print_timings:        eval time = 105441.63 ms /   127 runs   (  830.25 ms per run)
llama_print_timings:       total time = 142971.74 ms

**Comment by pjlegato** [comment]
The CMakeLists.txt seems to select CPU instruction set extensions manually (https://github.com/ggerganov/llama.cpp/blob/master/CMakeLists.txt#L54-L58 and https://github.com/ggerganov/llama.cpp/blob/master/CMakeLists.txt#L193-L219). This will cause the output binary to throw "illegal instruction" errors when run on a CPU that does not support the selected instruction sets.

Unless deliberately cross-compiling to generate a binary that will run on some other CPU (which is rare), the extension instruction sets should not be specified manually. Setting `-march=native` will automatically select the best instruction sets available on the current CPU at compile time.

More info at https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html

**Comment by dniku** [comment]
@pjlegato 
> Setting `-march=native` will automatically select the best instruction sets available on the current CPU at compile time.

That's theoretically true, but this whole thread is dedicated to the fact that when compiling the project on the Android device itself `-mcpu=native` (not `-march=native` though) causes `Illegal instruction` errors, while `-mcpu=generic` doesn't.

**Comment by dfyz** [comment]
@dniku

Oh my god, this proved to be a *deep* rabbit hole. I was able to reproduce this with a `Galaxy Z Flip4` (which uses `Snapdragon 8+ Gen 1`) and started investigating.

First off, the problematic instructions on your gdb screenshots (`cnth`, as you mentioned, but also `rdvl`) are part of [SVE](https://developer.arm.com/Architectures/Scalable%20Vector%20Extensions). So a more targeted workaround for this problem is replacing `-mcpu=native` with `-mcpu=native+nosve`, which prevents the compiler from trying to automatically vectorize loops using SVE. Fun fact: in this instance, clang decided to use SVE for [this loop](https://github.com/ggerganov/llama.cpp/blob/master/ggml.c#L3623) from `ggml_new_tensor_impl()` in order to speed up copying *[checks notes]* 2 (two) integers.

A much more interesting question is why this happens in the first place. As far as I understand, clang uses [this function](https://github.com/llvm/llvm-project/blob/fe1f34453d7ef2c0d5f308e0abdb185d6807cfa6/llvm/lib/TargetParser/Host.cpp#L159) to parse the first 32 lines of `/proc/cpuinfo` and determine the CPU model. More precisely, it uses the `CPU implementer` and `CPU part` fields. In our case, the implementer is `0x41` (ARM Ltd.) and the parts are `0xd46` (`cortex-a510`), `0xd47` (`cortex-a710`), and `0xd48` (`cortex-x2`). This is somewhat strange (my older Samsung has `0x53 -- Samsung Electronics Co., Ltd.` in the `CPU implementer`, which makes more sense), but apparently the newer Snapdragon Kryo CPUs are indeed [based](https://en.wikipedia.org/wiki/Kryo#8_series) on ARM cores.

My initial guess was that only some of the cores implemented SVE, but it turned out that all 3 Cortex models [are](https://github.com/llvm/llvm-project/blob/4889214a48cf7c7d1949b833d5a2d4604448c96e/llvm/include/llvm/TargetParser/AArch64TargetParser.h#L364) ARMv9, where both SVE and its extension SVE2 are mandatory. Every description of Snapdragon 8 I can find on the Internet proudly declares that it is ARMv9-A and supports SVE(2). However, there are no indications of that in the `Features` field of `/proc/cpuinfo`. I then tried reading the SVE-related bits directly from [ID_AA64PFR0_EL1](https://developer.arm.com/documentation/ddi0595/2021-06/AArch64-Registers/ID-AA64PFR0-EL1--AArch64-Processor-Feature-Register-0?lang=en#fieldset_0-35_32) with a simple C program:
```c
#include <stdio.h>
#include <inttypes.h>

int main() {
        uint64_t features = 0;
        asm("mrs %0, ID_AA64PFR0_EL1" : "=r"(features));
        printf("features: %lu\n", (features >> 32) & 0x1F);
}
```

It also prints 0 every time I run it. I don't believe it is possible for an OS to influence reads from this register, so it appears to be a genuine issue with the CPU. [This comment](https://github.com/android/ndk/discussions/1706#discussioncomment-4986595) also seems to confirm this.

So, to summarize: Qualcomm cores pretend to be ARMv9 cores while not being compliant ARMv9 cores, and clang believes them.

**Comment by dfyz** [comment]
@aicoat

Did you manage to resolve you problem? A52 seems to use an older Kryo core based on Cortex-A76, so the problem you were seeing might have been unrelated to the SVE one that @dniku and @GH228 stumbled upon.

If you still are seeing `Illegal instruction`, it would be great to run `main` under `gdb` (just as @dniku did) to see what the problematic instruction is.

**Comment by Topping1** [comment]
> I have found a solution. It might not be the best one, but it seems to work.
> 
> First, apply the following patch:
> 
> ```diff
> diff --git a/CMakeLists.txt b/CMakeLists.txt
> index 1a434f0..bd98b97 100644
> --- a/CMakeLists.txt
> +++ b/CMakeLists.txt
> @@ -186,7 +186,7 @@ if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm" OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES
>          # TODO: arm msvc?
>      else()
>          if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "aarch64")
> -            add_compile_options(-mcpu=native)
> +            add_compile_options(-mcpu=generic)
>          endif()
>          # TODO: armv6,7,8 version specific flags
>      endif()
> ```
> 
> Next,
> 
> ```shell
> mkdir build && \
> cd build && \
> cmake .. && \
> make main && \
> mv bin/main .. && \
> cd .. && \
> rm -r build
> ```
> 
> After this, `examples/chat.sh` works for me.
> 
> Funny enough, none of `{cortex-x2,cortex-a710,cortex-a510}` work. All three throw `Illegal instruction` errors.
> 
> By the way, this solution was suggested by ChatGPT:
> 
> ![1000008237.jpg](https://user-images.githubusercontent.com/1315874/229316464-92a12c99-60b2-4c2b-8595-99c427cb9a30.jpg)

I had the same problem compiling the code in my phone, a Xiaomi Mi 9 Se (Snapdragon 712 with 6Gb of Ram). Your edit to CMakeLists.txt fixed the issue.

**Comment by yifeifang** [comment]
-mcpu=generic worked for me

**Comment by thuningxu** [comment]
`-mcpu=native+nosve` works for me (Samsung S23 Ultra, Snapdragon 8 Gen 2)

**Comment by jo-elimu** [comment]
> i compiled the source using wsl2 with ndk r25 without any errors.

@aicoat How did you compile the source code on Android? Would you mind sharing the commands you used?

**Comment by jojorne** [comment]
LostRuins fixed it by removing the mcpu native:

```c++
ifneq ($(filter aarch64%,$(UNAME_M)),)
	# Apple M1, M2, etc.
	# Raspberry Pi 3, 4, Zero 2 (64-bit)
	# Nvidia Jetson
	MK_CFLAGS   +=
	MK_CXXFLAGS +=
```

**Comment by ajiekc905** [comment]
New tag b4358 feature detection os not working properly on Snapdragon 8 gen 1.
As a result - illegal instruction.
b4357 work without an issue.
